src/role_management.c

```diff
@@ -0,0 +1,722 @@
+ * to promote in our place. */
+static void handoverVoterCb(struct polling *polling)
+{
+	struct dqlite_node *node;
+	raft_id leader_id;
+	const char *borrowed_addr;
+	char *leader_addr;
+	struct compare_data voter_compare = {0};
+	unsigned i;
+	struct all_node_info *cluster;
+	unsigned n_cluster;
+	dqlite_node_id target_id;
+	struct handover_voter_data *data;
+	uv_work_t *work;
+	void (*cb)(struct dqlite_node *, int);
+	int rv;
+
+	if (polling == NULL) {
+		return;
+	}
+	node = polling->node;
+	cluster = polling->cluster;
+	n_cluster = polling->n_cluster;
+	cb = node->handover_done_cb;
+
+	raft_leader(&node->raft, &leader_id, &borrowed_addr);
+	if (leader_id == node->raft.id || leader_id == 0) {
+		goto finish;
+	}
+	leader_addr = raft_malloc(strlen(borrowed_addr) + 1);
+	if (leader_addr == NULL) {
+		goto finish;
+	}
+	memcpy(leader_addr, borrowed_addr, strlen(borrowed_addr) + 1);
+
+	/* Select a non-voter to transfer to -- the logic is similar to
+	 * adjustClusterCb. */
+	for (i = 0; i < n_cluster; i += 1) {
+		if (cluster[i].online && cluster[i].role == DQLITE_VOTER &&
+		    cluster[i].id != node->raft.id) {
+			addDomain(cluster[i].failure_domain, &voter_compare);
+		}
+	}
+	qsort_r(cluster, n_cluster, sizeof *cluster, compareNodesForPromotion,
+		&voter_compare);
+	target_id = 0;
+	for (i = 0; i < n_cluster; i += 1) {
+		if (cluster[i].online && cluster[i].role != DQLITE_VOTER &&
+		    cluster[i].id != node->raft.id) {
+			target_id = cluster[i].id;
+			break;
+		}
+	}
+	/* If no transfer candidates found, give up. */
+	if (target_id == 0) {
+		goto err_after_alloc_leader_addr;
+	}
+
+	/* Submit the handover work. */
+	data = raft_malloc(sizeof *data);
+	if (data == NULL) {
+		goto err_after_alloc_leader_addr;
+	}
+	data->node = node;
+	data->target_id = target_id;
+	data->leader_addr = leader_addr;
+	data->leader_id = leader_id;
+	work = raft_malloc(sizeof *work);
+	if (work == NULL) {
+		goto err_after_alloc_data;
+	}
+	work->data = data;
+	rv = uv_queue_work(&node->loop, work, handoverVoterWorkCb,
+			   handoverVoterAfterWorkCb);
+	if (rv != 0) {
+		goto err_after_alloc_work;
+	}
+	return;
+
+err_after_alloc_work:
+	raft_free(work);
+err_after_alloc_data:
+	raft_free(data);
+err_after_alloc_leader_addr:
+	raft_free(leader_addr);
+finish:
+	node->handover_done_cb = NULL;
+	cb(node, DQLITE_ERROR);
+}
+
+static void handoverTransferCb(struct raft_transfer *req)
+{
+	struct dqlite_node *d = req->data;
+	raft_free(req);
+	pollCluster(d, handoverVoterCb);
+}
+
+void RoleManagementAdjust(struct dqlite_node *d)
+{
+	/* Only the leader can assign roles. */
+	if (raft_state(&d->raft) != RAFT_LEADER) {
+		return;
+	}
+	/* If a series of role adjustments is already in progress, don't kick
+	 * off another one. */
+	if (!QUEUE__IS_EMPTY(&d->changes)) {
+		return;
+	}
+	/* If a handover is in progress or the node has shut down, do nothing.
+	 */
+	if (d->status != DQLITE_SERVER_RUNNING) {
```

Comment:
I think this can be turned into an assertion: `assert(d->status == DQLITE_SERVER_RUNNING`, or rather `assert(d->running)` considering the other comment. There should be no way that `RoleManagementAdjust()` can be called while the server is not running.

This will also eliminate the "line not covered by tests" warning emitted by Codecov below.